### PR TITLE
implement torch.addr using TensorIterator based kernels

### LINKED ISSUE #47313: addr bug for complex and inefficient implementation
`addr` implementation here 
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebra.cpp#L147-L159
does explicit multiplications by `alpha` and `beta`  even though by default `alpha` and `beta` are 1, so no multiplication is necessary. This degrades performance and memory usage. 
In addition, line 55 does invalid cast, so invocation with complex beta results in an error:
```
a=torch.randn(3, dtype=torch.complex64)                                                                                                                                                                
b=torch.randn(3, dtype=torch.complex64)                                                                                                                                                                
c=torch.randn(3,3, dtype=torch.complex64)   
c.addr(a,b, beta=0+0.1j)
```
```
RuntimeError: value cannot be converted to type double without overflow: (0,0.1)
```  
related: #44364
cc @vishwakftw @jianyuh @nikitaved @pearu @mruberry @heitorschueroff
cc @ivannz

### PR DESCRIPTION
**Pull request by RockingJavaBean**
Fixes #47313
Related #44364

This PR implements `torch.addr` function using `TensorIterator` with `cpu_kernel_vec` and `gpu_kernel`.
It helps reduce memory usage, improve performance, and fix the bug when `beta` or `alpha` is a complex number.

Todo
- [x] benchmarking `torch.addr` for the change of this PR, as well as the legacy TH implementation used in PyTorch 1.6.

### CODE REVIEW

**Comment by RockingJavaBean** [comment]
@JackCaoG This PR implements `torch.addr` using `TensorIterator` with corresponding cpu and cuda kernels.
The `test_broadcast_fused_matmul_xla` failed in the XLA CI job.
```
Nov 10 15:49:31 ERROR [3.464s]: test_broadcast_fused_matmul_xla (__main__.TestTorchDeviceTypeXLA)
Nov 10 15:49:31 ----------------------------------------------------------------------
Nov 10 15:49:31 Traceback (most recent call last):
Nov 10 15:49:31   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 272, in instantiated_test
Nov 10 15:49:31     result = test_fn(self, *args)
Nov 10 15:49:31   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 8451, in test_broadcast_fused_matmul
Nov 10 15:49:31     r0 = fntorch(t0_small, t1, t2)
Nov 10 15:49:31 RuntimeError: Cannot access data pointer of Tensor that doesn't have storage
```

Please kindly confirm whether I could skip the `torch.addr` XLA error with the below updates for now.
- remove the `addr` from the `fns` list
- add a new test (`test_broadcast_fused_addr`) with `onlyOnCPUAndCUDA` annotation

https://github.com/pytorch/pytorch/blob/ae5c2febb912066f1a8dec8b54451c09195b2c6d/test/test_torch.py#L8419-L8423

**Comment by RockingJavaBean** [comment]
The performances of `torch.addr` were benchmarked for the following versions of PyTorch.
- This PR: TensorIterator based kernels implementation
- PyTorch 1.8 nightly build: composite ops based implementation
- Pytorch 1.6.0: legacy TH code, implemented using Blas solutions

As `torch.utils.benchmark` were not introduced in PyTorch 1.6.0, I manually ported it by copying the corresponding codes to the installed path of PyTorch 1.6.0.

<details>
<summary>Benchmark script</summary>
<p>

```python
import torch
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
from itertools import product

results = []
num_threads = 1

size = 5000

beta = 1.2
alpha = 2.2

dtypes = torch.testing.get_all_fp_dtypes()
devices = [torch.device('cpu'), torch.device('cuda')]

for device, dtype in product(devices, dtypes):
    if torch.__version__ == '1.6.0' and device == torch.device('cuda') and dtype == torch.bfloat16:
        continue
    if torch.__version__ == '1.6.0' and device == torch.device('cpu') and dtype == torch.half:
        continue

    M = torch.randn(size, size).to(device=device, dtype=dtype)
    vec1 = torch.randn(size).to(device=device, dtype=dtype)
    vec2 = torch.randn(size).to(device=device, dtype=dtype)

    tasks = [("torch.addr(M, vec1, vec2, beta=beta, alpha=alpha)", f"{dtype}")]
    timers = [Timer(stmt=stmt,
                    num_threads=num_threads,
                    label=f"{torch.__version__}, size: {size}",
                    sub_label=f"{device}",
                    description=dtype,
                    globals=globals()) for stmt, dtype in tasks]

    repeats = 3
    for i, timer in enumerate(timers * repeats):
        results.append(timer.blocked_autorange())
        print(f"\r{device}-{dtype} {i + 1} / {len(timers) * repeats}")

comparison = Compare(results)
comparison.print()

```
</p>
</details>


**This PR**
```
[-------------------------- 1.8.0a0+9cad980, size: 5000 --------------------------]
            |  torch.float32  |  torch.float64  |  torch.float16  |  torch.bfloat16
1 threads: ------------------------------------------------------------------------
      cpu   |       65.9      |      164.4      |      909.9      |      190.7
      cuda  |        4.1      |        8.2      |        3.6      |        3.9

Times are in milliseconds (ms).
```

**PyTorch 1.8 nightly build**
```
[------------------------- 1.8.0.dev20201110, size: 5000 -------------------------]
            |  torch.float32  |  torch.float64  |  torch.float16  |  torch.bfloat16
1 threads: ------------------------------------------------------------------------
      cpu   |      191.8      |      382.4      |      1256.1     |      300.1
      cuda  |       16.9      |       32.4      |         9.9     |       10.2

Times are in milliseconds (ms).
```

**PyTorch 1.6.0**
```
[------------------------------- 1.6.0, size: 5000 -------------------------------]
            |  torch.float32  |  torch.float64  |  torch.bfloat16  |  torch.float16
1 threads: ------------------------------------------------------------------------
      cpu   |       59.0      |      120.8      |      152.6       |
      cuda  |       13.2      |       24.9      |                  |       5.5

Times are in milliseconds (ms).
```
The CPU path does not support `torch.float16`, while the CUDA path does not support `torch.bfloat16` in PyTorch 1.6.0.

**Result**
The TensorIterator based implementations outperformed both the composite ops and BLAS based implementions in most cases.
It was sightly slower than the BLAS ones in terms of `torch.float32` and `torch.float64` for the CPU path, while it increased the CUDA performance to a significant percentange.

Please kindly help review this PR. @ngimel @mruberry

File: test/test_linalg.py [hunk 1]
```diff
@@ -70,41 +67,84 @@ def run_test_case(m, a, b, beta=1, alpha=1):
...
+        with self.assertRaisesRegex(RuntimeError,
+                                    'Boolean alpha only supported for Boolean results.'):
+            self._test_addr_vs_numpy(device, dtype, beta=2, alpha=True)
+
+        # when beta is zero
+        self._test_addr_vs_numpy(device, dtype, beta=0., alpha=2)
+        # when beta is not zero
+        self._test_addr_vs_numpy(device, dtype, beta=0.5, alpha=2)
+        if dtype in torch.testing.get_all_complex_dtypes():
+            self._test_addr_vs_numpy(device, dtype, beta=(0 + 0.1j), alpha=(0.2 - 0.2j))
```
Comment (RockingJavaBean) [inline test/test_linalg.py:147]: This is to test the scenarios when `beta` or `alpha` is a complex number.

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -143,34 +147,85 @@ static void check_1d(const Tensor& t, const char* arg, const char* fn) {
...
 }
 
-Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const Tensor& vec2, Scalar beta, Scalar alpha) {
-  auto addr_result = at::addr(self, vec1, vec2, beta, alpha);
-  // Validates safe casting
-  const auto result_dtype = addr_result.scalar_type();
-  TORCH_CHECK(canCast(result_dtype, result.scalar_type()),
-              "result type ", result_dtype,
-              " can't be cast to the desired output type ", result.scalar_type());
+Tensor& addr_out(Tensor &result,
```
Comment (ailzhang) [inline aten/src/ATen/native/LinearAlgebra.cpp:219]: One difference made in this PR is the old implementation works for all backends but now it only support CPU and CUDA only. 
So here are a few missing changes in this PR:
1. change the addr.out entry in native_functions.yaml from DefaultBackend to CPU,CUDA only.  
2. Save the old implementation of addr_out in a math_addr_out function and put it as Math keyword in native_functions.yaml. see https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md#choosing-the-right-dispatch-keyword for more details. 
3. Add a test in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/test/math_kernel_test.cpp to make sure the new kernels you wrote for CPU & CUDA produces the same results as the old kernels we have. 
The second item should fix the xla test failure properly. 
All of the steps above should be smalll changes I think. 
Please let me know if you have followup questions! Thanks!

**Review by ailzhang — CHANGES_REQUESTED** [review]
Left some comments on the backend support. Please feel free to reach out if you have any questions!

File: aten/src/ATen/native/cpu/LinearAlgebraKernel.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                 Scalar beta, Scalar alpha) {
+  if (iter.dtype() == ScalarType::Bool) {
+    using scalar_t = bool;
+    auto beta_val = beta.to<scalar_t>();
+    auto alpha_val = alpha.to<scalar_t>();
+    cpu_kernel(iter,
+     [=](scalar_t self_val,
+         scalar_t vec1_val,
+         scalar_t vec2_val) __ubsan_ignore_undefined__ -> scalar_t {
+        return beta_val * self_val + alpha_val * vec1_val * vec2_val;
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/LinearAlgebraKernel.cpp:19]: this does not ignore `self_val` if beta=0, thus nans and infs in `self` won't be ignored, same in cuda kernel

File: test/test_linalg.py [hunk 4]
```diff
@@ -296,41 +293,83 @@ def run_test_case(m, a, b, beta=1, alpha=1):
...
-        run_test_case(m, zero_strided, b, beta=0.5, alpha=2)
+        zero_strided = make_tensor((1,), device=device, dtype=dtype, low=-2, high=2).expand(50)
+        check(m, zero_strided, b, beta, alpha)
 
         # test scalar
         m_scalar = torch.tensor(1, device=device, dtype=dtype)
-        run_test_case(m_scalar, a, b)
+        check(m_scalar, a, b, beta, alpha)
+
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
```
Comment (ngimel) [inline test/test_linalg.py:327]: nit: numpy is required, no need for this skipIf

**Review by ngimel — APPROVED** [review]
couple small comments, also, can you please rebase?

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -559,13 +559,7 @@ def merge_dicts(*dicts):
...
 
 For inputs of type `FloatTensor` or `DoubleTensor`, arguments :attr:`beta` and
-:attr:`alpha` must be real numbers, otherwise they should be integers
-
-.. warning::
-    This function is deprecated and may be removed in a future release.
-    It can be implemented using :func:`torch.outer` as
-    ``alpha * torch.outer(vec1, vec2) + beta * input`` when :attr:`beta` is not zero,
-    and as ``alpha * torch.outer(vec1, vec2)`` when :attr:`beta` is zero.
+:attr:`alpha` must be real numbers, otherwise they should be integers.
```
Comment (mruberry) [inline torch/_torch_docs.py:562]: This paragraph is out of date and the term "real numbers" is ambiguous.

**Review by mruberry — APPROVED** [review]
Overall looks good; I pointed out one docs issue.

File: torch/_torch_docs.py (hunk 5 above)
Comment (RockingJavaBean) [reply torch/_torch_docs.py]: Thanks so much for the suggestion, the paragraph has been removed.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 2 above)
Comment (RockingJavaBean) [reply aten/src/ATen/native/LinearAlgebra.cpp:219]: @ailzhang Thank you so much for your kind help in fixing the XLA failures, and providing the detailed docs and samples.
I have addressed the suggested changes and the XLA CI checks passed.

File: aten/src/ATen/native/cpu/LinearAlgebraKernel.cpp (hunk 3 above)
Comment (RockingJavaBean) [reply aten/src/ATen/native/cpu/LinearAlgebraKernel.cpp]: Thanks so much for pointing this out. I have updated the `CPU` and `cuda` kernels, the corresponding CPP and Pytorch test codes have been updated as well.

File: test/test_linalg.py (hunk 4 above)
Comment (RockingJavaBean) [reply test/test_linalg.py]: Addressed.

**Comment by RockingJavaBean** [comment]
@mruberry @ngimel Thanks so much for the review.
I have updated the PR according to the comments and rebased the code, `Circle CI` checks seem all passed except for one broken upstream CI.
Please kindly take a look.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**